lib/ddtrace/contrib/mongodb/subscribers.rb

```diff
@@ -0,0 +1,104 @@
+        def started(event)
+          pin = Datadog::Pin.get_from(event.address)
+          return unless pin && pin.enabled?
+
+          # start a trace and store it in the current thread; using the `operation_id`
+          # is safe since it's a unique id used to link events together. Also only one
+          # thread is involved in this execution so thread-local storage should be safe. Reference:
+          # https://github.com/mongodb/mongo-ruby-driver/blob/master/lib/mongo/monitoring.rb#L70
+          # https://github.com/mongodb/mongo-ruby-driver/blob/master/lib/mongo/monitoring/publishable.rb#L38-L56
+          span = pin.tracer.trace('mongo.cmd', service: pin.service, span_type: Datadog::Ext::Mongo::TYPE)
+          Thread.current[:datadog_mongo_span] = span
+
+          # common fields for all commands
+          command_name = event.command_name
+          collection = event.command[command_name]
+          span.set_tag(Datadog::Ext::Mongo::COLLECTION, collection)
+          span.set_tag(Datadog::Ext::Mongo::DB, event.database_name)
+          span.set_tag(Datadog::Ext::NET::TARGET_HOST, event.address.host)
+          span.set_tag(Datadog::Ext::NET::TARGET_PORT, event.address.port)
+
+          # commands are handled so that specific fields are normalized based on type, if it's
+          # a command that requires documents or a specific query. For some commands, we only
+          # take in consideration the query ('q') to keep the cardinality low
+          # NOTE: 'find' doesn't use a symbol
+          case command_name
+          when 'find'
+            filter = event.command['filter']
+            unless filter.nil? || filter.empty?
+              query = Datadog::Contrib::MongoDB.normalize_query(filter)
+              span.set_tag('mongodb.filter', query)
+            end
+          when :update
+            updates = event.command['updates']
+            unless updates.nil? || updates.empty? || updates.first['q'].empty?
+              query = Datadog::Contrib::MongoDB.normalize_query(updates.first['q'])
+              span.set_tag('mongodb.updates', query)
+            end
+          when :delete
+            deletes = event.command['deletes']
+            unless deletes.nil? || deletes.empty? || deletes.first['q'].empty?
+              query = Datadog::Contrib::MongoDB.normalize_query(deletes.first['q'])
+              span.set_tag('mongodb.deletes', query)
+            end
+          else
+            span.set_tag('mongodb.ordered', event.command['ordered'])
+
+            documents = event.command['documents']
+            unless documents.nil? || documents.empty?
+              document = Datadog::Contrib::MongoDB.normalize_documents(documents)
+              span.set_tag('mongodb.documents', document)
+            end
+          end
+
+          # set the resource with the quantized documents or queries
+          span.resource = "#{command_name} #{collection} #{document || query}".strip
```

Comment:
In `failed` and `succeeded` you have a rescue clause. I'm wondering if it applies here too. The question is: what happens if there's a typo in this code, is the consequence just a broken instrumentation or a problem in user code. The 1st is bad but acceptable, the 2nd is a strong no-go. You might want to check this, and add that `ensure` clause if it makes sense.

Reply:
The exception is handled but re-raised, so it's better to add a `rescue-ensure` statement here.

As in Python, `rescue` is not expensive so we can use it here. What is **really** expensive is if we hit the exception all the time, but this is definitely not the common case and we're expecting to never hit an exception here (otherwise we have a bug in our code). This is a safe choice and worth adding that one.